new/ohai-config.md

```diff
@@ -0,0 +1,89 @@
+
+## Specification
+
+
+`Ohai::Config` will use the new ChefConfig library that is bundled with the Chef client. Configuration will be set using a `Mixlib::Config` config_context named 'ohai'. Here is an example 'client.rb' file that would configure both the client and Ohai.
+
+```
+log_level        :info
+log_location     STDOUT
+chef_server_url  "https://api.chef.io/organizations/oneofus"
+ohai.plugin_path = "/etc/chef/ohai/plugins.local"
+ohai.plugin[:hostname][:fqdn_using] = [ :hostname, :nis, :dns ]
+```
+
+Because at the top level (outside of a config_context) `Ohai::Config` will be the same as `ChefConfig` and `Chef::Config`, existing top level configuration options like `Ohai::Config[:disabled_plugins]` will be deprecated in favor of new settings within the config_context, i.e. `Ohai::Config.ohai.disabled_plugins`. Until support for those top-level settings is removed, their values will be copied as appropriate for backward compatibility.
+
+For example:
+
+```
+# Old Syntax
+Ohai::Config[:disabled_plugins] = [ :BigPlugin ]
+
+# New Syntax
+Ohai::Config.ohai.disabled_plugins = [ :BigPlugin]
+```
+
+For convenience, a config method will be added to the `Ohai` class to access `Ohai::Config.ohai`, allowing the use of `Ohai.config[:plugin_path]` instead of `Ohai::Config.ohai[:plugin_path]` throughout the code. All existing uses of `Ohai::Config` in Ohai will need to be updated accordingly.
+
+### Configuration File
+
+When run from the command line, Ohai should load the workstation 'config.rb' and then the 'client.rb' files from the appropriate platform specific path, unless an alternate configuration file is provided as a command line argument. This provides similar behavior to knife which loads 'config.rb' and then 'knife.rb'. This facilitates reducing the number of separate configuration files to maintain for command line behavior.
+
+When loaded as a library, Ohai must not load a configuration file and will expect to be provided any necessary configuration options. For example, when loaded by the Chef client, configuration values will be located in the 'client.rb' file, or the file passed to the Chef client as the configuration file. Using the same file for configuration simplifies configuration file creation during bootstrap for the client.
+
+### Namespacing
+
+To reduce the risk of built-in and custom plugins using the same configuration setting for conflicting purposes, it is recommended that all plugins prefix their configuration settings with `[:plugin]` and the snake-case name of the plugin consuming the setting, as set in the plugin itself. The exposed overlap is intended, to facilitate passing a configuration option to the same plugin for multiple platforms but only specifying it once.
+
+For example, configuration file settings would look like:
+
+```
+ohai[:plugin][:memory][:unit] = "mb"
```

Comment:
From a technical perspective, this is a bit tricky. Allowing arbitrary nesting here means that we have to make `ohai[:plugin]` an auto-vivifying Hash or Mash. I dislike that because you have to fix the behavior of every single method on Hash to do the right thing (e.g., rarely used stuff like `Hash#replace`). Though this will be less of a problem if we limit the nesting level we allow for plugin configuration to a single level.